File: rename_xliff_files.py
#!/usr/bin/env python3

import os
import sys
import re
from pathlib import Path

def rename_xliff_files(directory='.'):
    """
    Rename XLIFF files based on the number in parentheses at the end of the filename.
    
    Examples:
    - translation_23925_de-du (983).xlf → 983.xlf
    - translation_23925_de-du (latest).xlf → LATEST.xlf
    """
    
    # Convert to Path object
    dir_path = Path(directory)
    
    # Check if directory exists
    if not dir_path.exists():
        print(f'Error: Directory "{directory}" does not exist')
        return
    
    print(f'\nScanning directory: {dir_path.resolve()}\n')
    
    # Find all XLIFF files
    xliff_files = list(dir_path.glob('*.xlf')) + list(dir_path.glob('*.xliff'))
    
    if not xliff_files:
        print('No XLIFF files found in the directory')
        return
    
    print(f'Found {len(xliff_files)} XLIFF file(s)\n')
    
    # Plan the renames
    renames = []
    
    for file_path in xliff_files:
        filename = file_path.name
        
        # Check for number in parentheses: (123)
        number_match = re.search(r'\((\d+)\)\.(xlf|xliff)$', filename, re.IGNORECASE)
        
        # Check for "latest" in parentheses: (latest)
        latest_match = re.search(r'\(latest\)\.(xlf|xliff)$', filename, re.IGNORECASE)
        
        new_name = None
        
        if number_match:
            # Use the number as the new filename
            number = number_match.group(1)
            extension = file_path.suffix
            new_name = f'{number}{extension}'
            
        elif latest_match:
            # Rename to LATEST
            extension = file_path.suffix
            new_name = f'LATEST{extension}'
        
        if new_name and new_name != filename:
            new_path = file_path.parent / new_name
            renames.append((file_path, new_path))
    
    if not renames:
        print('No files need renaming (no files with (number) or (latest) pattern found)')
        return
    
    # Show what will be renamed
    print('The following files will be renamed:\n')
    for old_path, new_path in renames:
        print(f'  {old_path.name}')
        print(f'  → {new_path.name}\n')
    
    # Ask for confirmation
    while True:
        answer = input('Do you want to proceed with renaming? (y/n): ').strip().lower()
        if answer in ['y', 'yes', 'n', 'no']:
            break
        print('Please answer y or n')
    
    if answer in ['y', 'yes']:
        # Perform the renames
        success_count = 0
        error_count = 0
        
        print()
        for old_path, new_path in renames:
            # Check if target already exists
            if new_path.exists():
                print(f'  ✗ Cannot rename "{old_path.name}" to "{new_path.name}" - file already exists')
                error_count += 1
                continue
            
            try:
                old_path.rename(new_path)
                print(f'  ✓ Renamed "{old_path.name}" to "{new_path.name}"')
                success_count += 1
            except Exception as e:
                print(f'  ✗ Error renaming "{old_path.name}": {e}')
                error_count += 1
        
        print(f'\nComplete: {success_count} file(s) renamed successfully')
        if error_count > 0:
            print(f'{error_count} file(s) failed to rename')
    else:
        print('Renaming cancelled')

def main():
    # Get directory from command line or use current directory
    directory = sys.argv[1] if len(sys.argv) > 1 else '.'
    rename_xliff_files(directory)

if __name__ == '__main__':
    main()

File: rename-xliff-files.js
#!/usr/bin/env node

const fs = require('fs');
const path = require('path');

// Get the directory from command line argument or use current directory
const directory = process.argv[2] || '.';

// Check if directory exists
if (!fs.existsSync(directory)) {
  console.error(`Directory "${directory}" does not exist`);
  process.exit(1);
}

console.log(`\nScanning directory: ${path.resolve(directory)}\n`);

// Read all files in the directory
const files = fs.readdirSync(directory);

// Filter for .xlf and .xliff files
const xliffFiles = files.filter(file => 
  file.endsWith('.xlf') || file.endsWith('.xliff')
);

if (xliffFiles.length === 0) {
  console.log('No XLIFF files found in the directory');
  process.exit(0);
}

console.log(`Found ${xliffFiles.length} XLIFF file(s)\n`);

// Process each file
const renames = [];
xliffFiles.forEach(file => {
  // Check for pattern with number in parentheses: (123)
  const numberMatch = file.match(/\((\d+)\)\.xlf/i);
  
  // Check for pattern with "latest" in parentheses: (latest)
  const latestMatch = file.match(/\(latest\)\.xlf/i);
  
  let newName = null;
  
  if (numberMatch) {
    // Extract the number and use it as the new filename
    const number = numberMatch[1];
    const extension = path.extname(file);
    newName = `${number}${extension}`;
  } else if (latestMatch) {
    // Rename to LATEST.xlf
    const extension = path.extname(file);
    newName = `LATEST${extension}`;
  }
  
  if (newName && newName !== file) {
    renames.push({ old: file, new: newName });
  }
});

if (renames.length === 0) {
  console.log('No files need renaming (no files with (number) or (latest) pattern found)');
  process.exit(0);
}

// Show what will be renamed
console.log('The following files will be renamed:\n');
renames.forEach(({ old: oldName, new: newName }) => {
  console.log(`  ${oldName}`);
  console.log(`  → ${newName}\n`);
});

// Ask for confirmation
const readline = require('readline');
const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout
});

rl.question('Do you want to proceed with renaming? (y/n): ', (answer) => {
  if (answer.toLowerCase() === 'y' || answer.toLowerCase() === 'yes') {
    // Perform the renames
    let successCount = 0;
    let errorCount = 0;
    
    renames.forEach(({ old: oldName, new: newName }) => {
      const oldPath = path.join(directory, oldName);
      const newPath = path.join(directory, newName);
      
      // Check if target file already exists
      if (fs.existsSync(newPath)) {
        console.error(`  ✗ Cannot rename "${oldName}" to "${newName}" - file already exists`);
        errorCount++;
        return;
      }
      
      try {
        fs.renameSync(oldPath, newPath);
        console.log(`  ✓ Renamed "${oldName}" to "${newName}"`);
        successCount++;
      } catch (error) {
        console.error(`  ✗ Error renaming "${oldName}": ${error.message}`);
        errorCount++;
      }
    });
    
    console.log(`\nComplete: ${successCount} file(s) renamed successfully`);
    if (errorCount > 0) {
      console.log(`${errorCount} file(s) failed to rename`);
    }
  } else {
    console.log('Renaming cancelled');
  }
  
  rl.close();
});